### DISCUSSION TITLE
StreamingResponse doesn't return until the first chunk of bytes are sent

### DISCUSSION THREAD

**Discussion by raishoemi** [fastapi/fastapi#6588] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request
import asyncio

app = FastAPI()

@app.get('/test_stream')
async def foo(request: Request):
    async def generator():
        await asyncio.sleep(10) # Simulate long running task to retrieve data
        yield 'data'
    return StreamingResponse(generator())
```

### Description

Hello,
I use `fastapi`'s `StreamingResponse` in order to return data to the user as soon as it's available.
I'm not sure why, but the response returns to the client only after the first chunk of data is sent. What I was hoping to get is 200 returned to the client when the `return StreamingResponse...` statement runs, and the data streamed will be sent once it's available.
Is this intended behavior? If so, is there a way to return a response immediately while streaming data later on?

Thanks!

### Operating System

Windows

### FastAPI Version

0.68.0

### Python Version

3.7.4

**Comment by ycd** [comment]
AFAICT this is expected behaviour.

**Comment by Kludex** [comment]
> The protocol server must not start sending the response to the client until it has received at least one Response Body event.

It's expected.

Reference: https://asgi.readthedocs.io/en/latest/specs/www.html#response-start-send-event

**Reply by pelson** [reply]
Worth noting that this spec now states:

> Protocol servers _need not_ flush the data generated by this event to the send buffer until the first Response Body event is processed

Context of this change is in https://github.com/django/asgiref/pull/389. (which comes from your issue raised in https://github.com/django/asgiref/pull/387).

In practice, I would like to be able to stream a response without doing any work until the client has started streaming the response. In my case, they may choose not to stream that response depending on the headers that were sent. At this moment, I don't think this is possible, but with that change in spec, it looks like it would be. Do you think this is a non-starter, or do you think I should open an issue on uvicorn to support the more relaxed specification?

**Comment by raishoemi** [comment]
I see, thanks for the help!
